app-typescript/components/WithPagination.tsx

```diff
@@ -0,0 +1,227 @@
+
+        if (lastInCurrentList !== totalPages - 1) { // add dots if we're skipping some numbers
+            basePages = basePages.concat('dots');
+        }
+
+        basePages = [
+            ...basePages,
+            totalPages,
+        ];
+    }
+
+    return basePages;
+}
+
+function getScrollParent(element: HTMLElement | null): HTMLElement | null {
+    if (element == null) {
+        return null;
+    }
+
+    let parentElement: HTMLElement | null = element;
+
+    const overflowY = window.getComputedStyle(parentElement).overflowY;
+    const hasScrollbar = overflowY === 'auto' || overflowY === 'scroll';
+
+    while (parentElement !== null && !hasScrollbar) {
+        parentElement = parentElement.parentElement ?? null;
+    }
+
+    return parentElement;
+}
+
+export class WithPagination<T> extends React.PureComponent<IProps<T>, IState<T>> {
+    private pageCount: number;
+    private abortController: AbortController;
+    private ref: HTMLDivElement | null;
+    private inProgress: boolean;
+
+    constructor(props: IProps<T>) {
+        super(props);
+
+        this.state = {
+            currentPage: 1,
+            items: null,
+        };
+
+        this.switchPage = this.switchPage.bind(this);
+        this.getPageSize = this.getPageSize.bind(this);
+
+        this.pageCount = 0;
+        this.abortController = new window.AbortController(); // window. needed for unit tests
+        this.ref = null;
+        this.inProgress = false;
+    }
+
+    getPageSize() {
+        return this.props.pageSize ?? 20;
+    }
+
+    switchPage(page: number) {
+        if (this.inProgress) {
+            this.abortController.abort();
+        }
+
+        this.inProgress = true;
+        this.props.getItems(page, this.abortController.signal).then((res) => {
+            this.inProgress = false;
+            this.setState({items: res.items, currentPage: page}, () => {
+                const scrollableEl = getScrollParent(this.ref);
+                const diff = scrollableEl != null && this.ref?.scrollHeight != null
+                    ? scrollableEl.offsetHeight - this.ref?.scrollHeight
+                    : null;
+
+                if (scrollableEl != null) {
+                    scrollableEl.scrollTop = diff != null ? diff : 0;
+                }
+            });
+        });
+    }
+
+    componentDidMount(): void {
+        this.props.getItems(1, this.abortController.signal).then((res) => {
+            this.pageCount = Math.ceil(res.itemCount / this.getPageSize());
+            this.setState({items: res.items});
+        });
+    }
+
+    render() {
+        if (this.state.items == null) {
+            return null;
+        }
+
+        const pageElements = getPagination(this.state.currentPage, this.pageCount).map((el, i) => {
+            if (el === 'dots') {
+                return (
+                    <span data-test-id="span" className='sd-pagination__item sd-pagination__item--more'>...</span>
```

Comment:
For data-test-id you should use names that make it easy to guess which UI element you're referring to.